File: rust/main/features/config/src/core/profile_loader.rs
use crate::api::{CleanProfile, ConfigProvider};

/// Load a CleanProfile from config by key.
pub fn load_profile(key: &str, config: &dyn ConfigProvider) -> CleanProfile {
    let prefix = format!("profiles.{key}");
    let get = |field: &str| config.get_value(&format!("{prefix}.{field}"));
    let get_arr = |field: &str| config.get_array(&format!("{prefix}.{field}"));

    CleanProfile {
        key: key.to_string(),
        name: get("name"),
        marker: get("marker"),
        alt_markers: get_arr("alt_markers"),
        profile_type: get("type"),
        command: get("command"),
        wrapper: get("wrapper"),
        wrapper_windows: get_arr("wrapper_windows").into_iter().next().unwrap_or_default(),
        clean_dir: get("clean_dir"),
        targets: get_arr("targets"),
        optional_targets: get_arr("optional_targets"),
        recursive_targets: get_arr("recursive_targets"),
        source_extensions: get_arr("source_extensions"),
        search_exclude: get_arr("search_exclude"),
        build_command: get("build_command"),
        output_pattern: get("output_pattern"),
    }
}

File: java/src/main/features/config/core/ProfileLoader.java
package main.features.config.core;

import main.features.config.api.CleanProfile;
import main.features.config.api.ConfigProvider;

/** Loads a CleanProfile from config by key. */
public class ProfileLoader {

    public static CleanProfile load(String key, ConfigProvider config) {
        var p = new CleanProfile();
        var prefix = "profiles." + key;

        p.key = key;
        p.name = config.getValue(prefix + ".name");
        p.marker = config.getValue(prefix + ".marker");
        p.altMarkers = config.getArray(prefix + ".alt_markers");
        p.profileType = config.getValue(prefix + ".type");
        p.command = config.getValue(prefix + ".command");
        p.wrapper = config.getValue(prefix + ".wrapper");
        p.wrapperWindows = config.getValue(prefix + ".wrapper_windows");
        p.cleanDir = config.getValue(prefix + ".clean_dir");
        p.targets = config.getArray(prefix + ".targets");
        p.optionalTargets = config.getArray(prefix + ".optional_targets");
        p.recursiveTargets = config.getArray(prefix + ".recursive_targets");
        p.sourceExtensions = config.getArray(prefix + ".source_extensions");
        p.searchExclude = config.getArray(prefix + ".search_exclude");
        p.buildCommand = config.getValue(prefix + ".build_command");
        p.outputPattern = config.getValue(prefix + ".output_pattern");

        return p;
    }
}
